### PR TITLE
Add survivalAFT base logic

### LINKED ISSUE #1557: Request: add survival regression
Would be nice if CatBoost implemented some form of survival / censored regression model, like [XGBoost](https://xgboost.readthedocs.io/en/latest/tutorials/aft_survival_analysis.html) and [Ranger](https://github.com/imbs-hl/ranger) do.

### PR DESCRIPTION
**Pull request by blatr**
Closes #1557
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=ru.

### CODE REVIEW

File: catboost/libs/helpers/distribution_helpers.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
+#include "distribution_helpers.h"
+
+EDistributionType DistributionFromString(const TString& distribution) {
```
Comment (ek-ak) [inline catboost/libs/helpers/distribution_helpers.cpp:3]: use function FromString

File: catboost/libs/helpers/distribution_helpers.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+double TExtremeDistribution::CalcPdfDer1(double pdf, double x) const {
+    const double expX = fast_exp(x);
+    return !IsFinite(expX) ? 0.0 : ((1 - expX) * pdf);
+}
+
+double TExtremeDistribution::CalcPdfDer2(double pdf, double x) const {
+    const double expX = fast_exp(x);
+    if (!IsFinite(expX) || !IsFinite(std::pow(expX, 2))) {
+      return 0.0;
```
Comment (ek-ak) [inline catboost/libs/helpers/distribution_helpers.cpp:76]: 4 spaces

File: catboost/libs/helpers/distribution_helpers.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    virtual double CalcPdf(double x) const = 0;
+    virtual double CalcCdf(double x) const = 0;
+    virtual double CalcPdfDer1(
+        double pdf,
+        double x) const = 0;
+    virtual double CalcPdfDer2(
+        double pdf,
+        double x) const = 0;
+    virtual EDistributionType GetDistributionType() const = 0;
+    virtual ~IDistribution()= default;
```
Comment (ek-ak) [inline catboost/libs/helpers/distribution_helpers.h:29]: space

File: catboost/libs/metrics/metric.cpp [hunk 4]
```diff
@@ -403,6 +403,71 @@ void TCtrFactorMetric::GetBestValue(EMetricBestValue* valueType, float* bestValu
     *bestValue = 1;
 }
 
+/* SurvivalAFT */
+namespace {
+    struct TSurvivalAftMetric final: public TAdditiveMultiRegressionMetric {
+        explicit TSurvivalAftMetric(const TLossParams& params)
+            : TAdditiveMultiRegressionMetric(ELossFunction::SurvivalAft, params){
```
Comment (ek-ak) [inline catboost/libs/metrics/metric.cpp:410]: ) {

File: catboost/private/libs/algo_helpers/error_functions.h [hunk 5]
```diff
@@ -10,6 +10,7 @@
 #include <catboost/private/libs/options/catboost_options.h>
 #include <catboost/private/libs/options/enums.h>
 #include <catboost/private/libs/options/restrictions.h>
+#include <catboost/private/libs/algo_helpers/survival_aft_utils.h>
```
Comment (ek-ak) [inline catboost/private/libs/algo_helpers/error_functions.h:13]: #include "survival_aft_utils.h" and higher

File: learn_error.tsv [hunk 6]
```diff
@@ -0,0 +1,101 @@
+iter	SurvivalAft:dist=Extreme
```
Comment (ek-ak) [inline learn_error.tsv:1]: catboost/learn_error.tsv is definitely wrong file path

**Review by ek-ak — CHANGES_REQUESTED** [review]

File: catboost/private/libs/algo_helpers/survival_aft_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,242 @@
+#include "survival_aft_utils.h"
+
+
+EDistributionType DistributionFromString(const TString& distribution){
```
[the remark this answers is not in the document]
Comment (blatr) [reply catboost/private/libs/algo_helpers/survival_aft_utils.cpp]: fixed

File: catboost/private/libs/algo/tensor_search_helpers.cpp [hunk 8]
```diff
@@ -145,6 +146,36 @@ THolder<IDerCalcer> BuildError(
...
+               switch(DistributionFromString(lossParams.at("dist"))){
+                   case EDistributionType::Extreme:
+                       distribution = std::make_unique<TExtremeDistribution>();     
+                       break; 
+                   case EDistributionType::Logistic:
+                       distribution = std::make_unique<TLogisticDistribution>();     
+                       break; 
+                   case EDistributionType::Normal:
+                       distribution = std::make_unique<TNormalDistribution>();     
+                       break;
```
[the remark this answers is not in the document]
Comment (blatr) [reply catboost/private/libs/algo/tensor_search_helpers.cpp]: fixed

File: catboost/libs/helpers/distribution_helpers.cpp (hunk 1 above)
Comment (blatr) [reply catboost/libs/helpers/distribution_helpers.cpp]: fixed

File: catboost/libs/helpers/distribution_helpers.cpp (hunk 2 above)
Comment (blatr) [reply catboost/libs/helpers/distribution_helpers.cpp]: fixed

File: catboost/libs/helpers/distribution_helpers.h (hunk 3 above)
Comment (blatr) [reply catboost/libs/helpers/distribution_helpers.h]: fixed

File: catboost/libs/metrics/metric.cpp (hunk 4 above)
Comment (blatr) [reply catboost/libs/metrics/metric.cpp]: fixed

File: catboost/private/libs/algo_helpers/error_functions.h (hunk 5 above)
Comment (blatr) [reply catboost/private/libs/algo_helpers/error_functions.h]: fixed

File: learn_error.tsv (hunk 6 above)
Comment (blatr) [reply learn_error.tsv:1]: fixed

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@ek-ak, internal review request updated: [1771007](https://a.yandex-team.ru/review/1771007)

**Pull request merged**